### PR TITLE
KAFKA-16318 : add javafoc for kafka metric

### PR DESCRIPTION
**Pull request by johnnychhsu**
## Context
There was no javadoc for `KafkaMetric`
Jira ticket: https://issues.apache.org/jira/browse/KAFKA-16318

## Solution
Add the javadoc for this class

## Test
Run './gradlew javadoc' and `./gradlew releaseTarGz` both succeeded.

Screenshot in the IDE
Screenshot 2024-03-07 at 12 20 04 AM

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 1]
```diff
@@ -20,6 +20,37 @@
...
+ * // set up metrics:
+ *
+ * Map<String, String> tags = new HashMap<>();
+ * tags.put("key1", "value1");
+ *
+ * MetricConfig config = new MetricConfig().tags(metricTags);
+ * Time time = new SystemTime();
+ * metricName = new MetricName(&quot;message-size-max&quot;, &quot;producer-metrics&quot;);
+ *
+ * KafkaMetric m = new KafkaMetric(new Object(),
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: It seems to me the point of usage is how to use its public methods rather than how to build this object. The reason of making `KafkaMetrics` be a public Api is that `MetricsReporter` (https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java#L39), which is a public interface, has many methods taking `KafkaMetric` as argument.

**Review by chia7712 — COMMENTED** [review]
@johnnychhsu thanks for taking over the jira. please take a look at the major comment.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 1 above)
Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: thanks for the comment! yes that makes sense to me, I think I missed the point.
Just updated and added doc for public methods, thanks!

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 2]
```diff
@@ -20,6 +20,37 @@
...
+ * An implementation of {@link Metric} interface.
+ * <p>
+ * A KafkaMetric is a named metric for monitoring purpose. The metric value can be a {@link Measurable} or a {@link Gauge}.
+ * <pre>
+ * <b>metricName</b> The name of the metric
+ * <b>lock</b> A lock used for reading the metric value in case of race condition
+ * <b>time</b> The POSIX time in milliseconds the metric is being taken
+ * <b>metricValueProvider</b> The metric collecting implementation that implements {@link MetricValueProvider}
+ * <b>config</b> The metric configuration which is a {@link MetricConfig}
+ * </pre>
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: This is already described in constructor, so I think we don't need it. WDYT?

**Comment by showuon** [comment]
@mimaison , please take a look when available. Thanks.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 3]
```diff
@@ -28,7 +59,14 @@ public final class KafkaMetric implements Metric {
     private final MetricValueProvider<?> metricValueProvider;
     private MetricConfig config;
 
-    // public for testing
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:31]: Let's not remove this comment. It's useful to indicate why this constructor is not private.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 2 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: Yeah since we don't expect users to create KafkaMetric objects I don't think we need this.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a couple of comments.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 2 above)
Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: thanks for the suggestion! let me remove that

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 3 above)
Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:31]: thanks for the comment! 
I see, let me add that back

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 4]
```diff
@@ -40,15 +48,28 @@ public KafkaMetric(Object lock, MetricName metricName, MetricValueProvider<?> va
         this.time = time;
     }
 
+    /**
+     * Get the configuration of this metric
+     * @return Return the config of this metric
+     */
     public MetricConfig config() {
         return this.config;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:57]: this is unrelate to this PR, but should we add lock to this method? It seems we assume the config can be read/write concurrently.

**Review by chia7712 — COMMENTED** [review]
@johnnychhsu nice patch. one minor comment left.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 4 above)
Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:57]: I think you are right, there is a lock for setter, we should have a lock for getter as well. 
Thanks for the suggestion, let me add that

**Comment by johnnychhsu** [comment]
tested with 
```
./gradlew cleanTest core:test --tests ConnectionQuotasTest --tests ControllerMutationQuotaTest stream:test --tests StandbyTaskTest --tests StreamTaskTest
```
and it passed

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 5]
```diff
@@ -40,15 +48,30 @@ public KafkaMetric(Object lock, MetricName metricName, MetricValueProvider<?> va
         this.time = time;
     }
 
+    /**
+     * Get the configuration of this metric
+     * @return Return the config of this metric
+     */
     public MetricConfig config() {
-        return this.config;
+        synchronized (lock) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: Could you please file another jira to trace it? This fix is good but it should be in another PR.

Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: sure, let me address this. I think you are right, this is a different issue. Thanks for pointing this out!

Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: this is addressed in [#15550](https://github.com/apache/kafka/pull/15550)

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 6]
```diff
@@ -78,6 +111,10 @@ public Measurable measurable() {
         }
     }
 
+    /**
+     * Set the metric config.
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:114]: This method normally is used to update quota and it should be used by server-side only. Could you add comments to remind `MetricsReporter` users that they should NOT call this method? ditto for `config()` method

Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:114]: I understand that for the config setter, users should not use it because it's aimed for server-side caller. However, why are users not allowed to fetch the config?

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:114]: > However, why are users not allowed to fetch the config?

`MetricConfig` has the methods used by server too, so encouraging user to access `MetricConfig` could lead potential trouble I feel.

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 7]
```diff
@@ -78,6 +109,11 @@ public Measurable measurable() {
...
+     * This is supposed to be used by server only, DO NOT use this for your metrics.
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java]: `This is supposed to be used by server only` this description is good enough to me

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java (hunk 6 above)
Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:114]: i see, thanks for the explanation. 
Just updated it, too.

**Review by chia7712 — APPROVED** [review]
LGTM

@mimaison @showuon please give a final review if you have free cycles. I will merge it tomorrow

**Comment by chia7712** [comment]
the failed tests are shown below, and they pass on my local
```script
./gradlew cleanTest :tools:test --tests ListConsumerGroupTest.testListConsumerGroupsWithTypesClassicProtocol --tests DescribeConsumerGroupTest.testDescribeStateWithMultiPartitionTopicAndMultipleConsumers :storage:test --tests TransactionsWithTieredStoreTest.testFencingOnAddPartitions --tests TransactionsWithTieredStoreTest.testCommitTransactionTimeout :connect:runtime:test --tests org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testAlterSinkConnectorOffsets --tests org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSourceConnector --tests org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testPotentialDeadlockWhenProducingToOffsetsTopic --tests org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSinkConnector --tests org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testTasksFailOnInabilityToFence --tests org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testGetSinkConnectorOffsetsDifferentKafkaClusterTargeted :metadata:test --tests QuorumControllerTest.testBootstrapZkMigrationRecord :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :server:test --tests ClientMetricsManagerTest.testCacheEviction :core:test --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeTopicAutoCreateTopicCreateAcl --tests AuthorizerIntegrationTest.shouldSendSuccessfullyWhenIdempotentAndHasCorrectACL --tests AuthorizerIntegrationTest.testAuthorizeByResourceTypeMultipleAddAndRemove --tests AuthorizerIntegrationTest.testConsumeWithTopicWrite --tests ControllerRegistrationManagerTest.testWrongIncarnationId --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests LogDirFailureTest.testIOExceptionDuringCheckpoint :clients:test --tests EagerConsumerCoordinatorTest.testOutdatedCoordinatorAssignment
```

File: clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java [hunk 8]
```diff
@@ -40,15 +48,29 @@ public KafkaMetric(Object lock, MetricName metricName, MetricValueProvider<?> va
         this.time = time;
     }
 
+    /**
+     * Get the configuration of this metric.
+     * This is supposed to be used by server only.
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:53]: What do you mean by `This is supposed to be used by server only.`?
Same for the other method below

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetric.java:53]: @mimaison it seems to me those methods are used by server to "write" something (for example, quota), and so users should view the metrics as readonly object. Please correct me If I misunderstand the purpose.

**Review by mimaison — APPROVED** [review]
